### ISSUE TITLE
Inputs and Outputs. 

### ISSUE THREAD

**Issue by vk2tds** [sanke/esp32_osdp_reader#1]
Alas, I am in lockdown away from my dev environment so developing is a bit tough at the moment. Great work on this. Looks much easier to use then Libosdp. 

The functionality I can see useful is, at least for me is:
* Inputs
* Outputs
* Modular multiple PD’s
* Un-linking Bluetooth and the OSDP

My use cases are:

* the wireless remote for my CP is about US$500 and the FOBs are not cheap. My old alarm has a good wireless unit with FOBs that look good. So, what I am thinking of doing is reverse engineering the comms on the receiver and sending the ID number as an ID. I would either have one virtual reader for each button, or one virtual reader with a virtual input for each button. I could haven have an automation that does things based on reader and input
* Another use case is a Wiegand to OSDP interface. Should be cheap
* yet another use case is multiple wiegand interfaces to OSDP

I look forward to getting access to hardware and hacking.

**Comment by sanke** [comment]
Hello, 

Thanks for your input. This project is/was more like a proof of concept for me. I was exploring fast physical authorization methods via BLE and the only convenient way to interface with existing access control devices was OSDP protocol (wiegand is another possibility but its going out of fashion due to lack of security). Beyond that I wasn't planning on developing this actively however writing C/C++ again was so much fun. I could spare some time and add some of your requested functionality in the future. Also feel free to contribute 

I want to clarify on modular multiple PD’s. Do I understand correctly that you want to expose several PD's via single physical connection?
All other points seems reasonable and clear (inputs, outputs and decoupling bluetooth from OSDP code).

**Comment by vk2tds** [comment]
Hi... I must apologise for the delay in replying. COVID has gotten in the way. I have started working through the code, and have found one interesting artifact where the ESP32 compiler does things differently from the 328P Arduino Compiler.

Consider the following code:
    `write(packet.get_data(), packet.get_len());`
On the ESP32, this presumably evaluates get_data() before get_len(). On the 328P, the get_len() is evaluated before the get_data(). In this case, the get_data() increases the size of the packet by adding a simple checksum. Therefore, the packet gets sent without the checksum. 

From what I can work out, this is undefined in the C++ spec. There used to be a compiler directive of 'pascal' to fix this on x86 platforms, but it has been a while, and I was brought up on 'C'

Multiple PD's are interesting since I could create a single device that implements four Wiegand to a single OSDP interface. Or have the four buttons on the remote FOB operate as two PD's. 

This is somewhat moot at the moment though. It appears the the InnerRange Inception requires encryption. 

```
RX 0x53 0x1 0x1B 0x0 0xE 0x3 0x13 0x0 0x77 0xDE 0xDC 0x93 0xB8 0xA0 0x89 0xAA 0x51 0x28 0x52 0x71 0xEF 0x61 0x18 0x8E 0x6E 0xEA 0x6E

RX 0x53 0x1 0x1E 0x0 0xD 0x2 0x17 0x80 0xBE 0xE9 0x93 0xA0 0xAD 0x68 0xC6 0x6E 0xB2 0x55 0xC2 0xBC 0x4E 0x81 0x3E 0xC3 0x95 0x72 0x7E 0x2 0xE4 0x14
 
```

I do note that [libosdp](https://github.com/goToMain/libosdp/tree/master/src/crypto) does have an AES library that might be of use. Alternately this might just push this project into a Too Hard basket which I solve with dollars rather than code. 

Once again, thanks for your time

**Comment by sanke** [comment]
Hello, no worries, having similar kind of issues as well. I started some work on supporting multiple PD's on a single comm channel this weekend. I plan to do some refactoring based on the image below

![image](https://user-images.githubusercontent.com/634568/136695457-88ece3de-9664-400b-a4d6-d899bbaf9f09.png)

My idea is that every OSDP PD has communication channel i.e. arduino serial interface and same channel can be reused across several PD's (I haven't studied libosdp, but I feel they are doing something similar to this and I'm reinventing the bike). This gives overall flexibility and encryption can be added seamlessly by switching comm channel implementation. What are your thoughts about this idea?

Regarding parameter evaluation order you are completely right. I think `get_data()` shouldn't mutate packet payload in the first place and a separate `write_checksum` method could be added so that evaluation order won't matter.

ATM I don't have hardware to test with, but I think I should be able to rig something up for testing using some existing implementation that support OSDP CP.

**Comment by vk2tds** [comment]
Thanks for the reply. I have managed to get libosdp compiled onto the Arduino toolchain, although I have not managed to get it working with my alarm system yet. I am using a STM32 Nucleo-64 board for this. Getting it running on a 328P seems to be memory related rather than flash. I suspect that by adjusting the buffers that it would be possible to get it working on the 328P, at least with a single PD with no capability to do advanced things like iris recognition :)

The two parts to the Arduino OSPD port of libospd are on my GitHub. But, it is a work in progress, and certain things are not working well at the moment, like the de-packetization. The Arduino has a serial buffer of up to 64 bytes, and the libospd assumes that a Serial Read will return less than a single packet. I have an in-progress routine to improve this, but it is deactivated at the moment for debugging. 

From what I have seen of the OSDP way of doing things, it seems that the encryption is per PD Context. So this needs to be abstracted out. The comms can be shared between multiple contexts, which appears to be the libospd way of doing things.

**Comment by sanke** [comment]
That's great! I'll use your port for my next relevant project if you don't mind. It's quite a challenge to make it run on 2KB of SRAM. Even OSDP protocol seems heavy in such context.

After tinkering at weekend I decided to back off from the idea of having my own lib :). I really forgot hardware constraint part (like you mentioned 328P), I thought high level code can run on anything these days. But it's not the case by far and I'm not into writing highly memory optimized code nor I have required experience and there is libosdp already.

Thanks for your replies and insights.

**Comment by vk2tds** [comment]
Thanks for the comments. I was quite amazed at the bang per buck you cab get on STM32 boards. I picked up some Nucleo-64 boards (F030R8 amd F411RE) and they were really cheap from RS/Element 14. As a bonus they have a detachable USB port that does FAST programming and debugging. I am using it in SWD mode if that helps. They have a heap more Serial, RAM and Flash than the 328P. And you can but cheap modules on AliExpress and eBay.

I have even found a couple of devices with RS485 built in, but they are not cheap as they have Ethernet on board.

The port is not yet working to the point where it will work with my alarm. It is acting as a PD sending packets at about 1 Hz though, so I think I am close. It is presently taking the other OSDP device offline on my alarm. Options for what is going on are either:
* Transmitting when it should not. I doubt that is happening
* Transmitting with a bad ID. But that is not happening
* Transmitting a packet that is causing the OSDP driver on my alarm to crash. 

The alarm in question is an InnerRange Inception, which I suspect is running Linux (or similar), so it wouldn't surprise me to find that the OSDP process is crashing. 

As for 2K... I have a 16k core memory board on the wall in my hallway. I am told when it was new, it was worth about the same as as a good new car. Crazy how things have changed

**Comment by sanke** [comment]
Yes, STM32 looks really capable MCU. Personally I really liked ESP-WROOM-32 it has wifi and bt integrated, good MCU and is relatively inexpensive.

For your problem I could suggest to sniff some comm's of working PD's and find what's different from yours. Oscilloscope with data packet analyzer could help as well

Yeah, I watch Dave Jones YouTube channel (EEVblog), from time to time he takes apart some old tech that was state of the art back in its day and was costing fortune. Really like those.
